opm/simulators/wells/GasLiftSingleWellGeneric.hpp

```diff
@@ -180,14 +181,20 @@ class GasLiftSingleWellGeneric
     std::pair<double, bool> getGasRateWithLimit_(const std::vector<double>& potentials) const;
     std::tuple<double,double,double,bool,bool,bool>
     getInitialRatesWithLimit_(const std::vector<double>& potentials);
+    std::tuple<double, const std::string*, double> getRateWithGroupLimit_(
+        Rate rate_type, const double new_rate, const double old_rate) const;
     std::pair<double, bool> getOilRateWithLimit_(const std::vector<double>& potentials) const;
     std::pair<double, bool> getWaterRateWithLimit_(const std::vector<double>& potentials) const;
 
-    std::pair<double, bool> getOilRateWithGroupLimit_(const double new_oil_rate, const double oil_rate) const;
-    std::pair<double, bool> getGasRateWithGroupLimit_(const double new_gas_rate, const double gas_rate) const;
-    std::pair<double, bool> getWaterRateWithGroupLimit_(const double new_water_rate, const double water_rate) const;
-    std::tuple<double,double,bool,bool> getLiquidRateWithGroupLimit_(double new_oil_rate, const double oil_rate,
-                                                                     double new_water_rate, const double water_rate) const;
+    std::pair<double, bool> getOilRateWithGroupLimit_(
+                           double new_oil_rate, double oil_rate) const;
+    std::pair<double, bool> getGasRateWithGroupLimit_(
+                           double new_gas_rate, double gas_rate) const;
+    std::pair<double, bool> getWaterRateWithGroupLimit_(
+                           double new_water_rate, double water_rate) const;
+    std::tuple<double,double,bool,bool> getLiquidRateWithGroupLimit_(
+                           double new_oil_rate, double oil_rate,
+                           double new_water_rate, double water_rate) const;
 
```

Comment:
Why are they made nonconst

Reply:
I could not see any reason to that these should be constant since they are not reference parameters. Why would you make them `const` here?

Reply:
> I could not see any reason why these should be constant since they are not reference parameters. Why would you make them `const` here?

On the other hand, if there's no reason to have mutable state&mdash;i.e., if you're actually changing the state&mdash;then then object should be declared `const` since that will generate build errors if you inadvertently change  something.

Reply:
> then the object should be declared const

@bska Yes I agree, but here we are passing simple types `double` which I would not call "object" and they are not passed by reference. On the other hand, if I had called `getLiquidRateWithGroupLimit_()` with a reference to an object of class `Foo` (for example) and I did not intend to modify the state of that object, I could declare the parameter as `const Foo& foo` to notify my intent to the compiler and humans reading the code. The compiler would then as a debugging aid check if I kept my promise.

Reply:
@hakonhagland the reasoning of @baska is not about preventing copies, but about preventing developers introducing bugs, see  [Cpp Core guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#S-const). But I have to admit that I usually do not adhere to this to (too much typing). I guess I need to step my game...